### PR TITLE
test: add fuzzer regression testing

### PR DESCRIPTION
**Pull request by keszybz**
This is a replacement for #7919.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -2572,6 +2589,46 @@ endforeach
 
 ############################################################
 
+prev = ''
+foreach p : fuzz_regression_tests
+        a = p.split('/')[-3]
+        b = p.split('/')[-2]
+        c = p.split('/')[-1]
```
Comment (titanous) [inline meson.build:2596]: `c` doesn't look used?

Comment (keszybz) [reply meson.build:2596]: It's used in line 2626.

Comment (titanous) [reply meson.build:2596]: :+1:

**Comment by yuwata** [comment]
With this PR, I got
```
Meson encountered an error in file meson.build, line 2590, column 16:
Unknown variable "sanitize_address".
```

**Comment by titanous** [comment]
I think moving `fuzz_regression_tests` into the above `if` statement will fix that. @yuwata it looks like it will work if you have `CC=clang CXX=clang` or similar.

File: meson.build [hunk 2]
```diff
@@ -2572,6 +2589,46 @@ endforeach
...
+
+        if a == 'address'
+                build = sanitize_address
```
Comment (yuwata) [inline meson.build:2599]: This fails when `Message: address sanitizer supported: no`.

**Comment by keszybz** [comment]
Thanks, updated.

**Comment by yuwata** [comment]
Thanks. Now this does not harm gcc builds.

File: meson.build [hunk 3]
```diff
@@ -2572,6 +2589,48 @@ endforeach
...
+                elif not sanitizers.contains(a)
+                        message('Not compiling @0@ because @1@ sanitizer is not available'.format(name, a))
+                elif slow_tests
+                        exe = custom_target(
+                                name,
+                                output : name,
+                                depends : build,
+                                command : [env,
+                                           'ln', '-fs',
+                                           join_paths(build.full_path(), b),
```
Comment (evverx) [inline meson.build]: While `meson-0.44.1` is able to successfully build `systemd` when `-Dslow-tests=true` is passed, `meson-0.42.1` fails with the following error:
```sh
Meson encountered an error in file meson.build, line 2619, column 16:
Command line arguments must be strings, files or targets.
```
. Should `meson_version` at the top of `meson.build` be bumped? Or could it be that the error is the result of something else?

Comment (keszybz) [reply meson.build]: Dunno. Can you debug it and come up with a version that works with old meson?

Comment (evverx) [reply meson.build]: `git bisect` says that the commit https://github.com/mesonbuild/meson/pull/2375/commits/bb0e18b73885de374f8461c0e4f3c911fded1e46 makes everything work, so the easiest solution would be to use `'meson_version : '>= 0.43'`. I'm not sure that I will be able to get around the bug in a timely manner, but probably @nirbheek might help.

Comment (keszybz) [reply meson.build]: That's tough. In Debian everything up to testing has meson-0.42. I'll try to make this work with older meson.

**Comment by keszybz** [comment]
Updated. Should now build with both old and new meson.

File: test/fuzz-regressions/meson.build [hunk 4]
```diff
@@ -0,0 +1,30 @@
+# SPDX-License-Identifier: LGPL-2.1+
+#
+# Copyright 2017 Zbigniew Jędrzejewski-Szmek
```
Comment (titanous) [inline test/fuzz-regressions/meson.build]: 2018

Comment (keszybz) [reply test/fuzz-regressions/meson.build]: Nitpicker! Updated.

**Comment by evverx** [comment]
@keszybz, thank you. I managed to build `systemd` with `meson-0.42.1`.

The PR looks good to me, but I'm a little bit puzzled about what should happen when `-Dinstall-tests=true` is used. In theory, it would be helpful if the fuzzers were installed along with the data.

**Comment by keszybz** [comment]
I don't think we can do that without some major hackery. The issue is that fuzzers use the sanitized version of libystemd-shared.so. In the build tree this is accomplished using rpath in the build tree. So we would need to install a second libsystemd-shared.so into some special path and make the installed fuzzers use this using rpath. In principle this can be done, but I'm not sure if it's worth the trouble. Apart from the complicated setup, this will pull in the sanitizer libraries as dependencies. I'd prefer no to do this for now, since the whole thing is very new and might change the way we do those tests as we get more cases.

**Comment by titanous** [comment]
@keszybz Don't forget to update the Semaphore config as per #7919 (with changes needed for this PR)

**Comment by keszybz** [comment]
Thanks for the reminder. It's done now.

**Comment by titanous** [comment]
@keszybz It looks like the build is broken due to these changes:

> add-apt-repository -y 'deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty main'
> Error: must run as root

**Comment by keszybz** [comment]
Pfff, after trying various combinations, I can't get clang-6.0 to install. We have some newer packages pulled in from artful because we needed something newer, and this creates version conflicts when I try to install clang-6.0. But I don't think clang-6.0 is actually _required_ for anything, since the fuzzers work with gcc / older clang. It's just not possible to do the fuzzing, but we don't need that. So I removed the changes for now. Maybe when semaphore migrates to a newer codebase, we can revisit this.

**Pull request merged**